contracts/impls/centralized/WitnetRequestBoard.sol

```diff
@@ -0,0 +1,278 @@
+
+
+    // ================================================================================================================
+    // --- Overrides 'Upgradable' -------------------------------------------------------------------------------------
+
+    /// @dev Initialize storage-context when invoked as delegatecall. 
+    /// @dev Should fail when trying to initialize same instance more than once.
+    function initialize(bytes memory _initData) virtual external override {
+        address _owner = __data().owner;
+        if (_owner == address(0)) {
+            // set owner if not yet set
+            _owner = msg.sender;
+            __data().owner = _owner;
+        }
+
+        // only owner can initialize:
+        require(msg.sender == _owner, "WitnetRequestBoard: only owner");
+
+        // current instance cannot be initialized more than once:
+        require(__data().instance != __stub, "WitnetRequestBoard: already initialized");
+        __data().instance = __stub;
+
+        // do actual initialization:
+        setReporters(abi.decode(_initData, (address[])));
+        emit Initialized(msg.sender, __stub);
+    }
+
+    /// @dev Retrieves named version of current implementation.
+    function version() virtual external pure override returns (string memory) {
+        return "WitnetRequestBoard-Centralized-v0.3.0";
+    }
+
+
+    // ================================================================================================================
+    // --- Utility functions not declared within an interface ---------------------------------------------------------
+
+    function setReporters(address[] memory reporters)
+        public
+        virtual
+        onlyOwner
+    {
+        for (uint ix = 0; ix < reporters.length; ix ++) {
+            address _reporter = reporters[ix];
+            if (_reporter != address(0)) {
+                __whitelists().isReporter_[_reporter] = true;
+            }
+        }
+    }
+
+    /// @dev Retrieves the whole DR post record from the WRB.
+    /// @param id The unique identifier of a previously posted data request.
+    /// @return The DR record.
+    function readDr(uint256 id)
+        external view
+        virtual
+        wasPosted(id)
+        returns (WitnetTypes.DataRequest memory)
+    {
+        return __checkDr(id);
+    }
+    
+    /// @dev Retrieves RADON bytecode of a previously posted DR.
+    /// @param id The unique identifier of the previously posted DR.
+    /// @return _bytecode The RADON bytecode. Fails if changed after being posted.
+    function readDataRequest(uint256 id)
+        external view
+        virtual
+        wasPosted(id)
+        returns (bytes memory _bytecode)
+    {
+        WitnetTypes.DataRequest storage _dr = __dataRequest(id);
+        _bytecode = WitnetRequest(_dr.addr).bytecode();
+        require(
+            WitnetTypes.computeDataRequestCodehash(_bytecode) == _dr.codehash,
+            "WitnetRequestBoard: manipulated request"
+        );
```

Comment:
Could this function be implemented as

```js
return readDr(id).bytecode();
```

or similar? Or it is more expensive?

Reply:
Actually, `readDr(uint)` returns a struct, not a contract. But yes, you are right if:
1.  `__checkDr(uint)` is changed as to check for bytecode integrity only if the DR was not deleted. 
2. `readDr(uint)` is declared as `public` instead of `external`.